crates/cli/src/subcommands/generate/csharp.rs

```diff
@@ -1,127 +1,533 @@
-        // Strings must have explicit default value of "".
-        AlgebraicType::String => Some(r#""""#),
-        AlgebraicType::Ref(r) => default_init(ctx, &ctx.typespace[*r]),
-        _ => {
-            debug_assert!(ty.is_scalar());
-            None
-        }
+            writeln!(
+                output,
+                "public SubscriptionBuilder<EventContext> SubscriptionBuilder() => new(this);"
+            );
+        });
+
+        vec![("SpacetimeDBClient.g.cs".to_owned(), output.into_inner())]
+    }
+
+    fn clap_value() -> clap::builder::PossibleValue {
+        clap::builder::PossibleValue::new("csharp").aliases(["c#", "cs"])
     }
 }
 
-// can maybe do something fancy with this in the future
-fn csharp_typename(ctx: &GenCtx, typeref: AlgebraicTypeRef) -> &str {
-    ctx.names[typeref.idx()].as_deref().expect("tuples should have names")
+fn ty_fmt<'a>(module: &'a ModuleDef, ty: &'a AlgebraicTypeUse) -> impl fmt::Display + 'a {
+    fmt_fn(move |f| match ty {
+        AlgebraicTypeUse::Identity => f.write_str("SpacetimeDB.Identity"),
+        AlgebraicTypeUse::Address => f.write_str("SpacetimeDB.Address"),
+        AlgebraicTypeUse::ScheduleAt => f.write_str("SpacetimeDB.ScheduleAt"),
+        AlgebraicTypeUse::Unit => f.write_str("SpacetimeDB.Unit"),
+        AlgebraicTypeUse::Option(inner_ty) => write!(f, "{}?", ty_fmt(module, inner_ty)),
+        AlgebraicTypeUse::Array(elem_ty) => write!(f, "System.Collections.Generic.List<{}>", ty_fmt(module, elem_ty)),
+        AlgebraicTypeUse::String => f.write_str("string"),
+        AlgebraicTypeUse::Ref(r) => f.write_str(&type_ref_name(module, *r)),
+        AlgebraicTypeUse::Primitive(prim) => f.write_str(match prim {
+            PrimitiveType::Bool => "bool",
+            PrimitiveType::I8 => "sbyte",
+            PrimitiveType::U8 => "byte",
+            PrimitiveType::I16 => "short",
+            PrimitiveType::U16 => "ushort",
+            PrimitiveType::I32 => "int",
+            PrimitiveType::U32 => "uint",
+            PrimitiveType::I64 => "long",
+            PrimitiveType::U64 => "ulong",
+            PrimitiveType::I128 => "I128",
+            PrimitiveType::U128 => "U128",
+            PrimitiveType::I256 => "I256",
+            PrimitiveType::U256 => "U256",
+            PrimitiveType::F32 => "float",
+            PrimitiveType::F64 => "double",
+        }),
+        AlgebraicTypeUse::Never => unimplemented!(),
+    })
 }
 
-macro_rules! indent_scope {
-    ($x:ident) => {
-        let mut $x = $x.indented(1);
-    };
+fn default_init(ctx: &TypespaceForGenerate, ty: &AlgebraicTypeUse) -> Option<&'static str> {
+    match ty {
+        // Options (`T?`) have a default value of null which is fine for us.
+        AlgebraicTypeUse::Option(_) => None,
+        AlgebraicTypeUse::Ref(r) => match &ctx[*r] {
+            // TODO: generate some proper default here (what would it be for tagged enums?).
```

Comment:
Alas, we have no way to translate `Default::default`.

...musing: we could add that information to the ModuleDef... not for here though.

Reply:
Yeah that could be nice.

Ideally, I'd omit this function altogether and mark all fields as `required` - which would force the user to fill them in explicitly, like Rust does - but, like many others, that's too modern C# feature not supported by Unity :(

Reply:
😭